descqa/StellarMassFunction.py

```
@@ -0,0 +1,353 @@
+            return TestResult(skipped=True, summary='Missing required quantity {} or {}'.format(self.zlabel, self.Mlabel))
+
+        #setup plots
+        fig, ax = plt.subplots(self.nrows, self.ncolumns, sharex='col')
+        fig.text(0.04, 0.5, self.yaxis, va='center', rotation='vertical') #setup a common axis label
+
+        #initialize arrays for storing histogram sums
+        N_array = np.zeros((self.nrows, self.ncolumns, len(self.Mbins)-1), dtype=np.int)
+        sumM_array = np.zeros((self.nrows, self.ncolumns,len(self.Mbins)-1))
+        sumM2_array = np.zeros((self.nrows, self.ncolumns,len(self.Mbins)-1))
+
+        #get catalog data by looping over data iterator (needed for large catalogs) and aggregate histograms
+        for catalog_data in catalog_instance.get_quantities([self.zlabel, self.Mlabel], filters=self.filters, return_iterator=True):
+            catalog_data = GCRQuery(*((np.isfinite, col) for col in catalog_data)).filter(catalog_data)
+            for cut_lo, cut_hi, N, sumM, sumM2 in zip_longest(
+                    self.z_lo,
+                    self.z_hi, 
+                    N_array.reshape(-1, N_array.shape[-1]), #flatten all but last dimension of array
+                    sumM_array.reshape(-1, sumM_array.shape[-1]),
+                    sumM2_array.reshape(-1, sumM2_array.shape[-1]),
+            ):
+                if cut_lo is not None:  #cut_lo can be 0. so cannot use if cut_lo
+                    mask = (cut_lo < catalog_data[self.zlabel]) & (catalog_data[self.zlabel] < cut_hi)
+                    M_this = catalog_data[self.Mlabel][mask]
+                    del mask
+
+                    #bin catalog_data and accumulate subplot histograms
+                    N += np.histogram(M_this, bins=self.Mbins)[0]
+                    sumM += np.histogram(M_this, bins=self.Mbins, weights=M_this)[0]
+                    sumM2 += np.histogram(M_this, bins=self.Mbins, weights=M_this**2)[0]
+
+        if not N.sum():
```

I understand what this is doing, but I wonder if we should raise an error instead of skip the catalog?

---

Yes, I agree. If the catalog claims to have a quantity but nothing is there (eg dc1) then we should know. Also the test condition was wrong, so I fixed it.